File: Data Structures & Algorithms/copy-linked-list-with-random-pointer/submission-2.py
"""
# Definition for a Node.
class Node:
    def __init__(self, x: int, next: 'Node' = None, random: 'Node' = None):
        self.val = int(x)
        self.next = next
        self.random = random
"""
class Solution:
    def copyRandomList(self, head: 'Optional[Node]') -> 'Optional[Node]':
        p = head
        map = {}
        while p :
            map[p]=Node(p.val)
            p =p.next
        p= head

        while p :
            copy=map[p]
            copy.next = map.get(p.next)
            copy.random=map.get(p.random)
            p=p.next
        return map.get(head)

File: Data Structures & Algorithms/linked-list-cycle-detection/submission-2.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

class Solution {
public:
    bool hasCycle(ListNode* head) {
         set<ListNode*> his;

    ListNode * curr = head;
    while (curr)
    {
        if (his.contains(curr))
            return true;

        his.insert(curr);
        curr=curr->next;
    }
    return false;
    }
};
